### PR TITLE
Add wrapper for Guzzle

### LINKED ISSUE #9054: Improve HTTPS cert validation for appstore
make code use https interface of appstore and properly validate it's certificate

cc @karlitschek

### PR DESCRIPTION
**Pull request by LukasReschke**
Other PR because Jenkins didn't like my old one :cry: 

<hr>

This adds a wrapper around Guzzle to allow developers to perform HTTP requests using a somewhat saner API. **Without this SSL and proxy support is basically broken in ownCloud as stated in https://github.com/owncloud/core/pull/14913#issuecomment-82490848**
## Todo
- [x] Get rid of private API
- [x] Deprecate HTTPHelper
- [x] Implement SSL in such a way that it actually works……
- [x] Follow proxy configured in config.php
- [x] Determine how to replace a users' custom trusted root certificate list
- [x] Fix unit tests after we agreed on an API design
- [x] Ensure that S2S still works properly

Requires https://github.com/owncloud/3rdparty/pull/167

Fixes https://github.com/owncloud/core/issues/9054, https://github.com/owncloud/core/issues/14963, https://github.com/owncloud/core/issues/14840 and a lot of other issues

<hr/>

Ref https://github.com/owncloud/core/pull/14913

### CODE REVIEW

**Comment by LukasReschke** [comment]
Ready to review now:

@nickvergessen @icewind1991 @MorrisJobke

**Comment by icewind1991** [comment]
Looks good :+1:

**Comment by MorrisJobke** [comment]
I tested to install an app and did a s2s sharing. There were no errors in the log.

**Comment by MorrisJobke** [comment]
Code also look good :+1:

**Comment by MorrisJobke** [comment]
3rdparty update: 69f14f9

File: apps/files_sharing/lib/controllers/sharecontroller.php [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 namespace OCA\Files_Sharing\Controllers;
 
+use Hoa\Core\Data\Data;
```
Comment (DeepDiver1975) [inline apps/files_sharing/lib/controllers/sharecontroller.php]: Hoa? WTF?

Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/sharecontroller.php]: Probably pressed enter when typing autocomplete for `DataResponse` :see_no_evil: 

… fixing …

File: lib/private/http/client/client.php [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+	private function setDefaultOptions() {
+		// Either use default bundle or the user bundle if nothing is specified
+		if($this->certificateManager->listCertificates() !== []) {
+			$dataDir = $this->config->getSystemValue('datadirectory');
+			$this->client->setDefaultOption('verify', $dataDir.'/'.$this->certificateManager->getCertificateBundle());
+		} else {
+			$this->client->setDefaultOption('verify', \OC::$SERVERROOT . '/config/ca-bundle.crt');
+		}
+
+		$this->client->setDefaultOption('headers/User-Agent', 'ownCloud Server Crawler');
```
Comment (DeepDiver1975) [inline lib/private/http/client/client.php]: what about firewalls rejecting anything which is not a browser? @danimo

**Pull request merged**